hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -1232,6 +1228,11 @@ def _draw_bounded_integer(
             # on other values we don't suddenly disappear when the gap shrinks to
             # zero - if that happens then often the data stream becomes misaligned
             # and we fail to shrink in cases where we really should be able to.
+            #
+            # TODO draw_bits isn't recorded in DataTree anymore, and soon won't
+            # have an impact on shrinking either. I think after shrinking is
+            # ported this hack can be removed, as we now always write the result
+            # of draw_integer to the observer, even when it's trivial.
```

Comment:
hmm, yes...this might also explain some shrink quality failures I've been seeing on this branch (`test_non_trivial_json`, `test_can_shrink_matrices_with_length_param`). We may want to revert to the "forcefully split single-valued nodes" approach and revisit this in the future.

Reply:
Agreed, let's keep the current (force-split) behavior for now.